scipy/spatial/ckdtree/src/build.cxx

```diff
@@ -112,53 +120,50 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
             std::nth_element(
                 node_indices, mid, node_indices + n_points, index_compare);
-
             split = data[*mid * m + d];
-            p = partition_pivot(node_indices, mid, split);
+            p = partition_pivot(indices + start_idx, indices + end_idx, split);
         }
         else {
             /* split with the sliding midpoint rule */
             split = (maxval + minval) / 2;
-
-            p = partition_pivot(indices + start_idx, indices + end_idx, split);
+            p = partition_pivot(indices + start_idx, indices + end_idx, split);            
         }
-
-        /* slide midpoint if necessary */
+        
+        // slide midpoint/pivot if necessary
+        // we might even need to do this for the median pivot to avoid infinite
+        // recursion          
         if (p == start_idx) {
             /* no points less than split */
             auto min_idx = *std::min_element(
-                indices + start_idx, indices + end_idx, index_compare);
-            split = std::nextafter(data[min_idx * m + d], HUGE_VAL);
+                    indices + start_idx, indices + end_idx, index_compare);
+            split = std::nextafter(data[min_idx * m + d], std::numeric_limits<double>::max() );
             p = partition_pivot(indices + start_idx, indices + end_idx, split);
-        }
-        else if (p == end_idx) {
+         }
+         else if (p == end_idx) {
             /* no points greater than split */
             auto max_idx = *std::max_element(
-                indices + start_idx, indices + end_idx, index_compare);
-            split = data[max_idx * m + d];
+                    indices + start_idx, indices + end_idx, index_compare);
+            split = std::nextafter(data[max_idx * m + d], std::numeric_limits<double>::lowest() );
             p = partition_pivot(indices + start_idx, indices + end_idx, split);
-        }
-
-        if (CKDTREE_UNLIKELY(p == start_idx || p == end_idx)) {
-            // All children are equal in this dimension, try again with new bounds
+         }
+          
+         if (CKDTREE_UNLIKELY(p == start_idx || p == end_idx)) {
+            // All children are equal in this dimension, try again with 
+            // this dimension tabooed
             assert(!_compact);
             self->tree_buffer->pop_back();
-            std::vector<double> tmp_bounds(2 * m);
-            double* tmp_mins = &tmp_bounds[0];
-            std::copy_n(mins, m, tmp_mins);
-            double* tmp_maxes = &tmp_bounds[m];
-            std::copy_n(maxes, m, tmp_maxes);
-
-            const auto fixed_val = data[indices[start_idx]*m + d];
-            tmp_mins[d] = fixed_val;
-            tmp_maxes[d] = fixed_val;
-
-            return build(self, start_idx, end_idx, tmp_maxes, tmp_mins, _median, _compact);
+            taboo[d] = true;
+            return build(self, start_idx, end_idx, maxes, mins, _median, _compact, taboo, taboo_depth+1);
```

Comment:
> I also find this strange code in the kd-tree construction [...] It tries to solve the issue of empty branches by pretending the spread along a dimension is zero.

Can you describe a situation where partitioning failed once already, but the data isn't fully degenerate?